src/ui/analyse/forecast/forecastView.ts

```diff
@@ -0,0 +1,164 @@
+import { playable } from '~/lichess/game'
+import { ForecastStep } from '~/lichess/interfaces/forecast'
+import settings from '~/settings'
+import { ontap } from '~/ui/helper'
+import AnalyseCtrl from '../AnalyseCtrl'
+import ForecastCtrl from './ForecastCtrl'
+import { groupMoves } from './util'
+
+type MaybeVNode = Mithril.Child | null
+
+function makeCandidateNodes(
+  ctrl: AnalyseCtrl,
+  fctrl: ForecastCtrl
+): ForecastStep[] {
+  const afterPly = ctrl.tree.getCurrentNodesAfterPly(
+    ctrl.nodeList,
+    ctrl.mainline,
+    ctrl.data.game.turns
+  )
+  return fctrl.truncate(
+    afterPly.map((node) => ({
+      ply: node.ply,
+      fen: node.fen,
+      uci: node.uci!,
+      san: node.san!,
+    }))
+  )
+}
+
+function renderNodesHtml(nodes: ForecastStep[]): MaybeVNode[] {
+  if (!nodes[0]) return []
+  if (!nodes[0].san) nodes = nodes.slice(1)
+  if (!nodes[0]) return []
+
+  return groupMoves(nodes).map(({ black, white, index }) => {
+    return h('move', [
+      h('index', index + (white ? '.' : '...')),
+      white ? h('san', white) : null,
+      black ? h('san', black) : null,
+    ])
+  })
+}
+
+function renderOnMyTurnView(ctrl: AnalyseCtrl, candidate: ForecastStep[]): MaybeVNode {
+  if (!ctrl.forecast?.isMyTurn) return
+  const firstNode = candidate[0]
+  if (!firstNode) return
+  const candidates = ctrl.forecast!.findStartingWithNode(firstNode)
+  if (!candidates.length) return
+
+  const lineCount = candidates.filter((candidate) => {
+    return candidate.length > 1
+  }).length
+
+  return (
+    h('div.on-my-turn',
+      h(
+        'button.defaultButton',
+        {
+          oncreate: ontap(() => ctrl.forecast!.playAndSave(firstNode))
+        }, [
+          h('span.fa.fa-check'),
+          h('span', [
+            h('strong', i18n('playX', candidate[0].san)),
+            ' ',
+            lineCount ? h('span', plural('andSaveNbPremoveLines', lineCount)) : null
+          ])
+        ],
+      )
+    )
+  )
+}
+
+function renderSpinner(): Mithril.Child {
+  return h('div.spinner_overlay', h('div.spinner.fa.fa-hourglass-half'))
+}
+
+export default function renderForecasts(ctrl: AnalyseCtrl): MaybeVNode {
+  const fctrl = ctrl.forecast
+  if (!fctrl || ctrl.synthetic || !playable(ctrl.data)) return null
+
+  const candidateNodes = makeCandidateNodes(ctrl, fctrl)
+  const isCandidate = fctrl.isCandidate(candidateNodes)
+
+  return (
+    h('div.analyse-training_box.analyse-forecast_box.box', {
+      className: fctrl.minimized ? 'minimized' : ''
+    }, [
+      renderTitle(fctrl),
+      h('div.forecasts-wrapper.native_scroller', [
+        h(
+          'div.forecasts-list',
+          {
+            class: settings.game.pieceNotation() ? 'displayPieces' : '',
+          },
+          [
+            ...fctrl.lines.map((nodes, i) => {
+              return h(
+                'div.forecast[data-icon=G]',
+                {
+                  key: `${i}-${nodes.map(node => node.san).join('')}`,
```

Comment:
I know, this basically defeats the purpose of keys... not proud of it. The issue is that if we just use the SAN, the ontap handler may have the wrong index if a forecast is deleted.

Ideally `ForecastCtrl` would maintain a map from SAN to `ForecastStep[]`, which would keep the key here constant and would let Mithril do its smart diffing magic.

Reply:
fixed!